Closes #1346: Add ReaderView feature component.

csadilek opened the pull request:
**DRAFT:** This brings in the Reader View feature component, implemented as a WebExtension. We need messaging support to finish this (see https://github.com/mozilla-mobile/android-components/issues/2243 and https://bugzilla.mozilla.org/show_bug.cgi?id=1518843), but for now this automatically activates reader view for `blog.mozilla.org`. 

Screenshots from our sample browser:

![readerview-demo](https://user-images.githubusercontent.com/472523/53903530-f3c66d00-4011-11e9-8dc2-4069def971c8.png)

Note that readability.js and readability-readerable.js are taken from: https://github.com/mozilla/readability

eeejay commented:
So awesome! Really excited to see this land.

This might have been discussed elsewhere, but we already bundle Firefox's reader mode in GeckoView (mostly by mistake), and their is an official [WebExtensions API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/toggleReaderMode) for it.

So ideally, this bundled extension could be made much simpler and not duplicate readability and all the functionality. And bugs.

The GeckoView WebExtensions support does not support the `tabs` namespace yet, so that's a blocker.

cc @agi90

csadilek commented:
> This might have been discussed elsewhere, but we already bundle Firefox's reader mode in GeckoView (mostly by mistake), and their is an official WebExtensions API for it.

@eeejay ah interesting! We did decide to go down this path and use the generic messaging API currently developed in GV, as we want tighter integration with the app e.g. we want to be able to have the font controls in the App UI (not part of content): https://github.com/mozilla-mobile/fenix/issues/190 (see mockups).

Theoretically, you / we can then remove this from GeckoView. Of course, we could re-think, but then we'd have to add reader-view-specific APIs and make it all customizable in GV. Readability is mostly brought in as a third-party lib here, but yes there is some duplication...

agi90 commented:
Opened https://bugzilla.mozilla.org/show_bug.cgi?id=1533156 to remove the code from GeckoView. Thank you @eeejay & @csadilek

pocmo commented on components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewFeature.kt:
```diff
@@ -0,0 +1,46 @@
+
+class ReaderViewFeature(
+    private val engine: Engine,
+    private val toolbar: Toolbar,
```
In the latest Fenix mocks "Reader View" is not a page action anymore but instead in this little drawer above the URL bar. So I assume we will need some kind of callback here so that the app can do something custom. I was actually wondering whether we should just expose this as a field on `Session`? This way we could have two feature classes, one installs the extension, updates the Session etc. And one the deals with the page action (which Fenix wouldn't need/use though!).

csadilek replied:
Good point, this wasn't thought out yet on my part. I've added a TODO to change to a callback (instead of passing in the toolbar), then the app can decide how to handle it. The idea of 2 feature classes sounds good as well, but I would like to try it out once we have messaging. Let's  chat more!

pocmo commented on components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewFeature.kt:
```diff
@@ -0,0 +1,46 @@
+                    Logger.error("Failed to install extension ${ext.id}", throwable)
+                }
+            )
+        }
```
Since this feature has activity/fragment scope and the engine "global" scope: What happens if this feature gets initialized multiple times (e.g. activity re-creation) and we install the extension again? Maybe that's another reason for separate features - a global one (that sets everything up) and a "local" one for messing with the UI?

agi90 replied:
yeah the web extension life cycle is tied to the runtime not the session.

csadilek replied:
True, there's already a check that makes sure we only install once, but we can split things up when we bring it all the pieces together.

pocmo commented on components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewFeature.kt:
```diff
@@ -0,0 +1,46 @@
+    private val engine: Engine,
+    private val toolbar: Toolbar,
+    sessionManager: SessionManager
+) : SelectionAwareSessionObserver(sessionManager), LifecycleAwareFeature {
```
nit: Extends `SelectionAwareSessionObserver` but doesn't use that yet? :)

csadilek replied:
Yes, for now it only registers the observer for the selected session, but there is nothing to observe yet :)

pocmo commented on samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt:
```diff
@@ -145,11 +146,14 @@ class BrowserFragment : Fragment(), BackHandler {
             owner = this,
             view = layout)
 
+        val readerViewFeature = ReaderViewFeature(components.engine, layout.toolbar, components.sessionManager)
```
nit: Since this keeps a strong reference to the View which may have a shorter lifetime than the fragment, we may want to wrap this in `ViewBoundFeatureWrapper`.

csadilek replied:
Ah good point. Fixed!

Amejia481 commented on components/feature/readerview/src/main/assets/extensions/readerview/moment/moment-2.24.0.js:
```diff
@@ -0,0 +1,14765 @@
+;(function (global, factory) {
```
We can use the [minified version](https://momentjs.com/downloads/moment.min.js) to save some bits or minify it our self

csadilek replied:
@Amejia481 yes, good idea! For now I want to be able to read / debug the code though. We can land the minified version or swap out in release builds.

csadilek replied:
Filed to investigate how we want to do this for release builds: https://github.com/mozilla-mobile/android-components/issues/2288

Amejia481 replied:
👍

Amejia481 commented on components/feature/readerview/src/main/assets/extensions/readerview/readability/readability-0.2.0.js:
```diff
@@ -0,0 +1,1800 @@
+/*eslint-env es6:false*/
```
This file could be minified too

csadilek replied:
https://github.com/mozilla-mobile/android-components/issues/2288

csadilek commented:
Now added JavaScript API to:
- set font size and adjust (increase/decrease) in steps (for the plus/minus buttons)
- set font type
- set color scheme (light|sepia|dark)

csadilek commented:
@pocmo I've updated this to sketch out the feature as discussed, and added TODOs where needed. 

I've also filed follow-ups:
- Engine API for messaging: https://github.com/mozilla-mobile/android-components/issues/2243
- Appearance UI: https://github.com/mozilla-mobile/android-components/issues/2623
- Wiring up messaging between Reader View feature and extension: https://github.com/mozilla-mobile/android-components/issues/2624

This is getting pretty big so I'd like to land it and work on the follow-ups :).

pocmo commented on components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewFeature.kt:
```diff
@@ -0,0 +1,141 @@
+            engine.installWebExtension(readerViewWebExt,
+                onSuccess = {
+                    Logger.debug("Installed extension: ${it.id}")
+                    installed = true
```
Since this callback is asynchronous, it is still technically possible to call `installWebExtension` twice here..?

csadilek replied:
OK, I've fixed this for now, but I think I will try to move the installation state into our `WebExtension` abstraction when we add more more functionality to it (as part of #2243)

csadilek commented on components/feature/readerview/src/main/assets/extensions/readerview/readerview.js:
```diff
@@ -0,0 +1,247 @@
+
```
Thanks! Added.

csadilek commented on components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewFeature.kt:
```diff
@@ -0,0 +1,141 @@
+    private val onReaderViewAvailableChange: OnReaderViewAvailableChange = { }
+) : SelectionAwareSessionObserver(sessionManager), LifecycleAwareFeature, BackHandler {
+
+    private val config = Config(PreferenceManager.getDefaultSharedPreferences(context))
```
OK. Done!

csadilek commented:
Addressed comment and added README.
